File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/models/tile.ts
import { Color } from '../../../core/models/color.model';
import { TileModel } from '../../../core/models/mosaic-project.model';
import { Vector } from '../../../core/models/vector.model';
import { calculatePointToPolygonDistance } from '../helpers/polygon-helpers/point-to-polygon-distance-calculator';

export class Tile implements TileModel {
    public id: string;

    public color: Color;
    public vertices: Vector[];

    public outerRadius: number;
    public innerRadius: number;

    public constructor(tile: TileModel) {
        this.id = tile.id;
        this.color = tile.color;
        this.vertices = tile.vertices;

        this.outerRadius = Math.sqrt(Math.max(...this.vertices.map((v) => v.x * v.x + v.y * v.y)));
        this.innerRadius = calculatePointToPolygonDistance(Vector.zero, this.vertices);
    }
}

File: backend/Irrtirer/Irrtirer.Generator/Tiles/Tile.cs
using Irrtirer.Generator.Helpers.PolygonHelpers;
using System;
using System.Drawing;
using System.Linq;
using System.Numerics;

namespace Irrtirer.Generator.Tiles
{
    [Obsolete("Później zrobić prywatny")]
    public class Tile
    {
        public Guid Id { get; private set; }

        public Color Color { get; private set; }
        public Vector2[] Vertices { get; }

        public float OuterRadius { get; protected set; }
        public float InnerRadius { get; protected set; }

        public Tile(Guid id, Vector2[] unnormalizedVertices, Color color)
        {
            Id = id;

            this.Color = color;
            Vector2 centroid = PolygonHelper.CalculatePolygonCentroid(unnormalizedVertices);
            this.Vertices = unnormalizedVertices.Select(v => v - centroid).ToArray();
            this.OuterRadius = MathF.Sqrt(Vertices.Max(v => (v.X * v.X) + (v.Y * v.Y)));
            this.InnerRadius = PointToPolygonDistanceCalculator.CalculatePointToPolygonDistance(Vector2.Zero, this.Vertices);
        }

        public override bool Equals(object obj)
        {
            return obj is Tile tile && Id.Equals(tile.Id);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Id);
        }
    }
}
